Add support for channel mentions in webhook attachments

sbishel opened the pull request:
#### Summary
Implements client-side dynamic resolution for channel mentions in posts, enabling channel mentions (~channel-name) to render as clickable links in webhook attachments while fixing several long-standing stale data issues. In order to display as link, user was required to be a member. 

**Completely circumvents todays solution.** Currently when a user adds a channel mention. A post property is added to the post containing information to add the proper link to a channel mention for a "non" member user. 

**Please see attachment document attachment for soluion comparison.**
[CHANNEL_MENTIONS_SOLUTION_COMPARISON.md](https://github.com/user-attachments/files/23056250/CHANNEL_MENTIONS_SOLUTION_COMPARISON.md)

** Full analysis document attached to JIRA ticket ** 

Problems Solved
  1. Channel Mentions in Webhook Attachments (Primary Fix)
Previously, channel mentions only worked in the main message text but not in attachment fields. When integrations or bots included channel references in attachment content (pretext, text, field values), they rendered as plain text instead of clickable links for non-members. 

  2. Stale Data Issues (Critical Fixes)
The old implementation captured channel data at post creation time in post.props.channel_mentions, which never updated and caused multiple problems:
  - Channel visibility changes:
    - Private → Public: Old posts still showed plain text (confusing)
    - Public → Private: Old posts still showed clickable links ⚠️ SECURITY CONCERN - information disclosure
  - Channel renames: Old posts showed outdated channel names
  - Channel archives: Old posts still showed as active links (not sure about intent for this.)

  Solution: Client-Side Dynamic Resolution
  Instead of relying on stale server metadata, the webapp now:
  1. Extracts channel mentions from entire post (message + attachments) at render time
  2. Fetches missing channels dynamically via API (with Redux caching)
  4. Builds channelNamesMap from current Redux store state
  5. Always reflects real-time channel state (visibility, names, membership)

  Benefits

  ✅ Real-time accuracy - Channel mentions always reflect current state, not post creation time
  ✅ Self-healing - All historical posts automatically benefit without migration
  ✅ Security - Public→Private changes immediately hide links
  ✅ Consistency - All posts (old and new) work identically
  ✅ Webapp-only - No server changes

  Changes

  Client-side:
  - Convert PostMarkdown to functional component with React hooks
  - Add extractChannelMentionsFromPost() utility to scan message + attachments
  - Dynamically fetch missing channels using getChannelByNameAndTeamName()
  - Build channelNamesMap from Redux store with current channel data
  - Update unit tests for Redux-based resolution

  Behavior:
  - ✅ Channel mentions render as links in attachments (pretext, text, field values)
  - ✅ Channel mentions remain plain text in titles/labels (correct)
  - ✅ Private channels show as plain text for non-members
  - ✅ Channel state changes reflect immediately in all posts

  ###Performance Overhead

  Double Parsing:
  - Content is parsed twice with the same regex pattern:
    a. First in extractChannelMentionsFromPost() to determine which channels to fetch
    b. Second in autolinkChannelMentions() during Markdown rendering to convert ~channel-name to links
  - Impact is negligible for typical posts (regex matching is fast)
  - Could be optimized in future by extracting positions during first parse

  Network Traffic:
  - Additional API calls to getChannelByNameAndTeamName for channels not in Redux store
  - Each unique mentioned channel requires one API call per user session
  - Mitigations:
    - Redux caching: Each channel fetched only once, then reused across all posts
    - Most channels already in Redux from sidebar/recent activity
    - Only public channels can be fetched (private return 404, no retry)
    - Batching: All missing channels in a post fetched simultaneously

  User Experience

  Initial Render Delay:
  - Brief moment where channel mentions show as plain text before becoming links
  - Links appear after fetch completes (typically < 100ms)
  - Graceful degradation: Post is readable immediately, links enhance progressively

  Testing:
  - 8 new Playwright E2E tests covering all attachment scenarios
  - Comprehensive unit tests for channel mention extraction
  - All tests passing (E2E + unit)

#### Ticket Link
Fixes https://mattermost.atlassian.net/browse/MM-66266

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->
```release-note
Fix for correctly displaying Channel Mentions as Channel Links.
```

mattermost-build commented:
Error trying to update the PR.
Please do it manually.

hmhealey commented on webapp/channels/src/components/post_markdown/post_markdown.tsx:
```diff
@@ -53,106 +57,203 @@ export type OwnProps = {
 
-        // Proxy images if we have an image proxy and the server hasn't already rewritten the this.props.post's image URLs.
-        const proxyImages = !this.props.post || !this.props.post.message_source || this.props.post.message === this.props.post.message_source;
-        const channelNamesMap = isChannelNamesMap(this.props.post?.props?.channel_mentions) ? this.props.post?.props?.channel_mentions : undefined;
```
We can't fully get rid of the usage of this prop because it's used to support cross-team channel links in DMs. It's not fully reliable in doing that, but if we remove its usage entirely, it'll break what little support for that that we have. If we want to make these use more up-to-date names after a channels is renamed, we'll have to merge that data with the data in the Redux store while taking into account which team the channels are in

hmhealey commented on webapp/channels/src/components/post_markdown/post_markdown.tsx:
```diff
@@ -53,106 +57,203 @@ export type OwnProps = {
-            if (o && o.hook && this.props.post) {
-                message = o.hook(this.props.post, message);
+        // Dispatch fetch for each missing channel
+        mentionsToFetch.forEach((channelName) => {
```
This is moving a lot of logic into the body of the component than we should really have here. That's going to introduce a lot of optential for performance problems because this component will now re-render any time anything changes on any channel (the previous code looks like it'll only re-render for changes to the current channel) which includes when any user posts in it, and it'll re-run all the Markdown formatting as well when that happens, even if nothing that affects the text has changed.

As much as possible of this should be done in a separate Redux action to make it so that the component just needs to dispatch the action when it gets a new post without knowing anything else about goes into checking if the channels exist or not. It should be like `getMentionsAndStatusesForPosts` in `packages/mattermost-redux/src/actions/posts` where it encapsulates all of that logic. In particular, take a look at the `getNeededAtMentionedUsernamesAndGroups` function that's a part of that since it handles looking through all the different fields of message attachments to load the contents of those.

Overall, the end result of the fetching should ideally reduce the amount that actually takes place in the component to something like this with the rest of the logic encapsulated away in the store.
```js
useEffect(() => {
    dispatch(fetchChannelsForPostIfNeeded(post));
}, [dispatch, post]);
```

hmhealey commented on webapp/channels/src/components/post_markdown/post_markdown.tsx:
```diff
@@ -53,106 +57,203 @@ export type OwnProps = {
-                message = o.hook(this.props.post, message);
+        // Dispatch fetch for each missing channel
+        mentionsToFetch.forEach((channelName) => {
+            dispatch(getChannelByNameAndTeamName(postTeam.name, channelName, true) as any).catch((error: Error) => {
```
Note that this will result in potentially many requests being made by the app at once, especially if you switch to a channel with many of these posts in it. Ideally, we would add an API to fetch multiple channels by name at once to avoid that, but that can be added in the future.

hmhealey commented:
Generally speaking, I do agree we should move more to the client. The server-side fix is needed for the case where we can't infer which channel a team belongs on (currently just posts in DMs/GMs and cross-team search results, but we've also talked about adding more support for cross-team channel shortlinks), but we shouldn't need it in the general case of a user just posting in a channel. I'm actually surprised that we never implemented something like that before since I can't remember us loading the channels referenced by a post on the fly.

This will still be different than how emojis and at-mentions are handled because those are pre-calculated by the server and included in the post metadata. There's some benefits to doing that (like avoiding followup requests from the client in response to a post and avoiding layout shift caused by the text changing as things load), but adding more to post metadata can slow down the createPost and getPosts APIs which are performance sensitive. Also, because of permissions, it could be a lot more work to treat channel shortlinks the same way as emojis and at-mentions where everyone can see them. Post metadata is nice for most cases, but it can only be used for posts, so you still need client-side handling in places that use Markdown that are outside of posts (like in the channel header).

All that said, there are some tradeoffs for doing this clientside (most notably the potential for followup requests after a new post is made and the potential for layout shift as mentioned above), but I think this is a good approach for this

sbishel commented:
Hey Harrison,

Thanks for the detailed feedback. I've implemented the server-side approach discussed above.

  1. Server-side scanning of attachments:
  - Added ChannelMentionsFromAttachments() to scan pretext, text, and field values
  - Extended FillInPostProps to use ChannelMentionsAll() (message + attachments)
  - Files: server/public/model/channel_mentions.go, server/public/model/post.go, server/channels/app/post.go

  2. Current team ID parameter for DMs:
  - Client sends current_team_id in post props for DM/GM posts only
  - Server uses it to scope channel search: teamId = channel.TeamId || currentTeamId
  - Removes it after use (transient hint, not persisted)
  - Fixes the cross-team disclosure security issue you identified

  3. Sanitization on retrieval (not creation):
  - Added sanitizeChannelMentionsForUser() called from SanitizePostListMetadataForUser
  - Split into two functions:
    - SanitizePostMetadataForUser - Creation (embeds only)
    - SanitizePostMetadataForUserOnRetrieval - GET endpoints (embeds + channel mentions)
  - Fetches fresh channel data and checks viewer permissions
  - Handles channel renames with fresh display names from DB

  4. Minimal client-side changes (~70 lines):
  - Prop threading through 3 attachment components
  - Sends current_team_id (15 lines in createPost)
  - Cross-team navigation fix (reads data-channel-mention-team attribute)
  - Selector reactivity fix (watches Redux post state for updates)

  ---
  📋 Addresses All 3 Problems

  | Problem                   | Solution                                                                      |
  |---------------------------|-------------------------------------------------------------------------------|
  | 1. Webhook attachments    | ✅ Server scans attachments, client threads channelNamesMap through components |
  | 2. Cross-team DMs         | ✅ current_team_id makes resolution deterministic, prevents disclosure         |
  | 3. Permission enforcement | ✅ Creator permissions at creation + viewer sanitization at retrieval          |

hmhealey commented on server/channels/app/post_metadata.go:
```diff
@@ -260,31 +260,112 @@ func (a *App) sanitizePostMetadataForUserAndChannel(rctx request.CTX, post *mode
+
+	mentionsMap, ok := channelMentionsProp.(map[string]any)
+	if !ok {
+		return post
```
I was going to suggest deleting the prop here when it's invalid, but I don't think that's actually possible since `App.FillInPostProps` already overwrites or deletes it. There's nothing to change here, but I wanted to mention it for posterity

hmhealey commented on webapp/channels/src/components/markdown/index.ts:
```diff
@@ -25,9 +25,15 @@ function makeGetChannelNamesMap() {
+        (channelNamesMap, propsChannelMentions, postChannelMentions) => {
+            // Use channel_mentions from post.props as source of truth (most up-to-date)
+            // This ensures selector re-runs when post.props.channel_mentions changes
+            const channelMentions = postChannelMentions || propsChannelMentions;
```
The `channelNamesMap` prop used by this component should just be `post.props.channel_mentions` already [as provided by the `PostMarkdown` component](https://github.com/mattermost/mattermost/blob/master/webapp/channels/src/components/post_markdown/post_markdown.tsx#L116), so I think we can remove `propsChannelMentions` and just rely on `postChannelMentions` here to simplify things

hmhealey reviewed (changes requested):
There's one other place we need to sanitize the channel mentions: in the WebSocket code. When a post is created or updated, post metadata is only generated once and `sanitizeChannelMentionsForUser` is just called once for the author before that post with metadata is passed to the WebSocket code for broadcasting to other users, so we need to strip out any mentions of private channels which the recipient of the message is not a part of.

For permalinks, that's done using a "broadcast hook" that's run by the server for each user just before the WS message is sent out, so we'll need one of those for channel mentions as well. It should be similar to the `permalinkBroadcastHook` which removes the props, so here's some links to the code for that:
- Where it's defined: https://github.com/mattermost/mattermost/blob/master/server/channels/app/web_broadcast_hooks.go#L161
- Where it's used: https://github.com/mattermost/mattermost/blob/master/server/channels/app/post.go#L920

hmhealey commented on server/channels/app/post.go:
```diff
@@ -1044,6 +1057,38 @@ func (a *App) setupBroadcastHookForPermalink(rctx request.CTX, post *model.Post,
+	}
+
+	// To remain secure by default, we wipe out the channel mentions unconditionally
+	post.DelProp(model.PostPropsChannelMentions)
```
Actually, `setupBroadcastHookForPermalink` removes the prop from the post at the start and then re-adds it at the end if the post's author has permissions to see the permalinked content. I think the "unconditional" part of that was because we wanted to sanitize the post whenever an error was encountered. Perhaps that's why that cloning wasn't necessary before since the post wasn't mutated in a way that mattered?

hmhealey commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -181,6 +183,85 @@ func (h *permalinkBroadcastHook) Process(msg *platform.HookedWebSocketEvent, web
+	}
+
+	var post model.Post
+	err = json.Unmarshal([]byte(currentPostJSON), &post)
```
Could this use `getTypedArg`? That covers lines 208 to 218 of this function, and it has additional error handling for if the `post` is missing from the WS message somehow

sbishel commented on server/channels/app/notification.go:
```diff
@@ -723,7 +723,10 @@ func (a *App) SendNotifications(rctx request.CTX, post *model.Post, team *model.
-	appErr := a.publishWebsocketEventForPost(rctx, post, message)
+	// Clone the post before publishing to WebSocket to avoid modifying the original post
+	// publishWebsocketEventForPost modifies the post in-place to set up broadcast hooks
+	postForBroadcast := post.Clone()
```
Removed and reworked.

sbishel commented on server/channels/app/post.go:
```diff
@@ -1044,6 +1057,38 @@ func (a *App) setupBroadcastHookForPermalink(rctx request.CTX, post *model.Post,
+		)
+		return model.NewAppError("setupBroadcastHookForChannelMentions", "app.post.marshal.app_error", nil, "", http.StatusInternalServerError).Wrap(err)
+	}
+	message.Add("post", postWithoutChannelMentionsJSON)
```
reworked

sbishel commented on server/channels/app/web_broadcast_hooks.go:
```diff
@@ -181,6 +183,85 @@ func (h *permalinkBroadcastHook) Process(msg *platform.HookedWebSocketEvent, web
+	})
+}
+
+type channelMentionsBroadcastHook struct{}
```
Done

hmhealey reviewed (approved):
Thanks for changing that around! It took a bit for me to trace my way through the updated code again since this part of the code is performance and security sensitive, but I think everything looks good now
